content/lessons/2_ownership/index.md

```diff
@@ -0,0 +1,346 @@
++++
+title = "Ownership Model"
+date = 2022-02-18
+weight = 1
+[extra]
+lesson_date = 2022-03-08
++++
+
+## Why all the fuss?
+
+Even if you've never seen Rust code before, chances are you still heard the term _borrow checker_ or something about Rust's ownership. Indeed, Rust's ownership model lies at the very core of its uniqueness. But to fully understand it and appreciate it, let's first take a look at how memory management is handled in most popular languages.
+
+- **Garbage Collection** - in many high-level programming languages, like Java, Haskell or Python, memory management is done fully by the language, relieving the programmer from this burden. This prevents memory leaks and memory related errors (like _use after free_), but does come at a cost - there is a runtime overhead, both memory and performance wise, caused by the constantly running garbage collection algorithms and the programmer usually has very little control over when the garbage collection takes place.
+
+- **Mind your own memory** - in low-level languages and specific ones like C++, performance comes first so we cannot really afford to run expansive bookkeeping and cleaning algorithms. Most of these languages compile directly to machine code and have no language-specific runtime environment. That means that the only place where memory management can happen is in the resulting code itself. While compilers insert these allocation / deallocation patterns in easy to spot places (like C++'s automatic management of stack allocated memory), it generally requires a lot of discipline from the programmer to adhere to good practices and patterns to avoid as many memory related issues as possible and one such bug can be quite deadly to the program and a nightmare to find and fix. These languages basically live by the _"your memory, your problem"_ mantra.
+
+And then we have Rust. Rust is a systems programming language and in many ways it's akin to C++ - it's basically low-level with many high-level goodies. But unlike C++, it doesn't exactly fall into either of these categories, though it's way closer to the second one. It performs no additional management at runtime, but instead imposes a set of rules on the code, making it easier to reason about and thus check for its safety and correctness at compile time - these rules make up Rust's **ownership model**.
```

Comment:
> into either of these categories, though it's way closer to the second one.

What categories? Hihg-level goodies is not a category

Reply:
"Categories" refers to the two approaches to memory management (either fully done by the language or fully expected from the programmer). But I can see how this wording is misleading, I can change it to something along the lines of:
_But unlike C++, it doesn't exactly fall into either of the categories described above..._